src/overlays/actors/ovl_Bg_Haka/z_bg_haka.c

```diff
@@ -32,21 +32,146 @@ const ActorInit Bg_Haka_InitVars = {
     (ActorFunc)BgHaka_Update,
     (ActorFunc)BgHaka_Draw,
 };
-*/
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Haka/BgHaka_Init.s")
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Haka/BgHaka_Destroy.s")
+static InitChainEntry sInitChain[] = {
+    ICHAIN_F32(minVelocityY, 0, ICHAIN_CONTINUE),
+    ICHAIN_VEC3F_DIV1000(scale, 100, ICHAIN_STOP),
+};
+
+extern UNK_TYPE D_06000428;
+extern Gfx D_060001B0[];
+extern Gfx D_060002A8[];
+
+void BgHaka_Init(Actor* thisx, GlobalContext* globalCtx) {
+    BgHaka* this = THIS;
+    s32 pad;
+    s32 sp24 = 0;
+
+    Actor_ProcessInitChain(&this->dyna.actor, sInitChain);
+    DynaPolyInfo_SetActorMove(&this->dyna, 0);
+    DynaPolyInfo_Alloc(&D_06000428, &sp24);
+    this->dyna.dynaPolyId = DynaPolyInfo_RegisterActor(globalCtx, &globalCtx->colCtx.dyna, &this->dyna.actor, sp24);
+    this->actionFunc = func_8087B7E8;
+}
+
+void BgHaka_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+    BgHaka* this = THIS;
+
+    DynaPolyInfo_Free(globalCtx, &globalCtx->colCtx.dyna, this->dyna.dynaPolyId);
+}
+
+void func_8087B758(BgHaka* this, Player* player) {
+    Vec3f sp1C;
+
+    func_8002DBD0(&this->dyna.actor, &sp1C, &player->actor.posRot.pos);
+    if (fabsf(sp1C.x) < 34.6f && sp1C.z > -112.8f && sp1C.z < -36.0f) {
+        player->stateFlags2 |= 0x200;
+    }
+}
+
+void func_8087B7E8(BgHaka* this, GlobalContext* globalCtx) {
+    Player* player = PLAYER;
+
+    if (this->dyna.unk_150 != 0.0f) {
+        if (globalCtx->sceneNum == SCENE_SPOT02 && LINK_IS_CHILD && gSaveContext.nightFlag == 0) {
+            this->dyna.unk_150 = 0.0f;
+            player->stateFlags2 &= ~0x10;
+            if (!Gameplay_InCsMode(globalCtx)) {
+                func_8010B680(globalCtx, 0x5073, NULL);
+                this->dyna.actor.params = 100;
+                this->actionFunc = func_8087BAE4;
+            }
+        } else if (0.0f < this->dyna.unk_150 ||
+                   (globalCtx->sceneNum == SCENE_SPOT06 && LINK_IS_CHILD && !Flags_GetSwitch(globalCtx, 0x23))) {
+            this->dyna.unk_150 = 0.0f;
+            player->stateFlags2 &= ~0x10;
+        } else {
+            this->dyna.actor.posRot.rot.y = this->dyna.actor.shape.rot.y + 0x8000;
+            this->actionFunc = func_8087B938;
+        }
+    }
+    func_8087B758(this, player);
+}
+
+void func_8087B938(BgHaka* this, GlobalContext* globalCtx) {
+    Player* player = PLAYER;
+    s32 sp38;
+    f32 temp_f0;
+
+    this->dyna.actor.speedXZ += 0.05f;
+    temp_f0 = this->dyna.actor.speedXZ;
+    if (temp_f0 > 1.5f) {
+        this->dyna.actor.speedXZ = 1.5f;
+    } else {
+        this->dyna.actor.speedXZ = temp_f0;
+    }
```

Comment:
this should be CLAMP_MAX(this->dyna.actor.speedXZ, 1.5f) and should hopefully get rid of the temp.
if the temp is still needed you can name it speedXZ or something